src/Constraint/Constraint.php

```diff
@@ -212,4 +220,96 @@ public function __toString()
     {
         return self::$transOpInt[$this->operator] . ' ' . $this->version;
     }
+
+    /**
+     * @inheritDoc
+     */
+    public function getLowerBound()
+    {
+        $this->extractBounds();
+
+        return $this->lowerBound;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function getUpperBound()
+    {
+        $this->extractBounds();
+
+        return $this->upperBound;
+    }
+
+    /**
+     * Takes a given constraint and a given version and creates
+     * the version_compare() statements required to evaluate
+     * whether a version matches upper and lower bounds.
+     *
+     * @param string $version
+     */
+    public static function getVersionCompareCondition(ConstraintInterface $constraint, $version)
```

Comment:
@Seldaek @naderman Doesn't really belong here and should go to its own class maybe. But how do we name it? And is the `VersionParser` injected or do we expect the `$version` to be normalized when passing?
It needs to be normalized because `version_compare('1.0', '1.0.0.0', '>=')` is `false`:(

Reply:
Hm that's a bit of a problem, we don't want to be normalizing PHP versions in our autoloader use case?

Reply:
Jup but I cannot change the way PHP works in `version_compare()` 😄

Reply:
And yes, the returned string should have `%s` and so it can inject the version at runtime and obviously not pre-generate it like it is now. Lol. But still, the main issue remains.

Reply:
I thought this was the point of the stuff at https://github.com/composer/semver/pull/75/files#diff-d094db696fbeed03e5e71f05bdb50ceeR282-R286 that it can be replicated easily at runtime without needing full normalizer.. Then could just pad with .0 until we have 4 dots?

Reply:
Normalizing just does what PHP does internally in `version_compare()`. See https://github.com/php/php-src/blob/6f9248359e5a0f9056e198a7ad9f87ca6e78aa56/ext/standard/versioning.c#L44-L46

Reply:
As for a class, I don't know call it Composer\Semver\ComparisonDumper or smth? :) 

And would be good to only dump a single version_compare if one of the bounds is infinite

Reply:
What about the:

> Then could just pad with .0 until we have 4 dots?

Reply:
The problem is not that semver needs to pad. The one comparing to it would need to pad.
You would need to normalize `PHP_VERSION` at runtime.

Reply:
In fact I could remove that part now that we're not calculating bounds ourselves anymore.

Reply:
I realize that, I'm saying the generated code should pad the $version/%s so that it matches the normalized/padded constraints. Kinda do a best effort minimal version of the normalization that hopefully is enough to make version_compare work.

Reply:
I got rid of the canonicalization stuff and implemented a ComparisonDumper which is also optimized to not dump lower and upper most bounds (so if both are given, it just dumps `true`).